Add minimal documentation for PEC in the Getting Started (#532)
unitaryfund/mitiq: docs/source/guide/guide-getting-started.rst

Note that we don't need to even redefine factories for different stacks. Once
you have a ``Factory`` it can be used with different front and backends.

Error Mitigation with Probabilistic Error Cancellation
------------------------------------------------------

> In *Mitiq*, it is very easy to switch between different error mitigation methods.

Reviewer: At some point we really need to decide on a convention for writing Mitiq in text: mitiq or Mitiq or *mitiq* or *Mitiq*? The docs are horribly inconsistent. It probably doesn't matter to anyone but me, but would be nice.

Reviewer: FWIW, I vote for Mitiq - capitalized everywhere, no special text.

Author: I completely agree! We should converge to a consistent notation in all the docs and README files.
I like the plain and capitalized Mitiq as you suggested.  If I am not wrong this is also what we used in the arXiv paper. 

I unified the notation only in this Getting Started for the moment, since this PR is about this section.

Reviewer: Nice! Yes it is "Mitiq" everywhere in the paper. Thanks for this.

For example, we can implement Probabilistic Error Cancellation (PEC) by using the same execution function (``executor``)
and the same *Cirq* circuit (``circuit``) that we have already defined in the section
:ref:`Multi-platform Framework <multi_platform_framework>`.

> Differently from ZNE, PEC requires the knowledge of the noise model and of the noise strength acting on the system.

Reviewer: I would say PEC depends on knowlege of the noise model and strength *more than ZNE*.

Author: I replaced this with:
`PEC requires a good knowledge of the noise model and of the noise strength acting on the system.`
without any comparison with ZNE.

In particular for each operation of the circuit, we need to build a quasi-probability representation of the 

> ideal unitary gate expanded in a basis of noisy implementable operations. For more details on PEC see
> the :ref:`Probabilistic Error Cancellation <guide_qem_pec>` section.

Reviewer: ```suggestion
ideal unitary gate expanded in a basis of noisy implementable operations. For more details behind the theory of PEC see
the :ref:`Probabilistic Error Cancellation <guide_qem_pec>` section.
```

In our simple case, ``circuit`` corresponds to the repetition of the same X gate,
whose representation in the presence of depolarizing noise can be obtained as follows:

.. testcode::

>     from mitiq.pec.representations import represent_operation_with_local_depolarizing_noise

Reviewer: ```suggestion
    from mitiq.pec import represent_operation_with_local_depolarizing_noise
```

    x_representation = represent_operation_with_local_depolarizing_noise(
        ideal_operation=Circuit(X(qubit)), 
        noise_level=NOISE_LEVEL,
    )

    print(x_representation)

.. testoutput::

    0: ───X─── = 1.010*0: ───X───-0.003*0: ───X───X───-0.003*0: ───X───Y───-0.003*0: ───X───Z───

The result above is an :class:`OperationRepresentation` object which contains the information for 
representing the ideal operation X (left-hand-side of the printed output) as a linear combination of
noisy operations (right-hand-side of the printed output). 

> We can now implement PEC by importing the function ``execute_with_pec`` from the 
> ``mitiq.pec`` module.

Reviewer 2: ```suggestion
We can now implement PEC by importing the function :func:`~mitiq.pec.pec.execute_with_pec` from the
:mod:`~mitiq.pec.pec` module.
```

Author: Thanks Nathan I always forget about this! Now I did this for all the functions/methods/classes of the Getting Started section

.. testcode::

    from mitiq.pec import execute_with_pec

    SEED = 0
    exact_result = 1
    noisy_result = executor(circuit)

>     pec_result = execute_with_pec(circuit, executor, [x_representation], random_state=SEED)

Reviewer: ```suggestion
    pec_result = execute_with_pec(circuit, executor, representations=[x_representation], random_state=SEED)
```

    print(f"Error without mitigation: {abs(exact_result - noisy_result):.{3}}")
    print(f"Error with mitigation (PEC): {abs(exact_result - pec_result):.{3}}")
    
    ratio = abs((exact_result - noisy_result) / (exact_result - pec_result))
    print(f"Mitigation (PEC) provides a {ratio:.{3}} factor of improvement.")

.. testoutput::

    Error without mitigation: 0.0387
    Error with mitigation (PEC): 0.00364

>     Mitigation (PEC) provides a 10.6 factor of improvement.

Reviewer: I worry about quoting these factors of improvement. It's too easy for someone who doesn't know what's going on to say "Only a 10x improvement? ZNE got a 167x improvement, it must be way better."

Author: I agree, these are just toy examples and quantitative improvements are quite meaningless for a 6 gate circuit. So I removed them from both the ZNE and the PEC results.
